File: library/src/main/java/tw/shounenwind/plurkconnection/callbacks/BasePlurkCallback.java
package tw.shounenwind.plurkconnection.callbacks;

import java.net.HttpURLConnection;

import okhttp3.Response;
import tw.shounenwind.plurkconnection.BuildablePlurkConnection;
import tw.shounenwind.plurkconnection.NewThreadRetryExecutor;
import tw.shounenwind.plurkconnection.PlurkConnectionException;
import tw.shounenwind.plurkconnection.responses.ApiResponseString;
import tw.shounenwind.plurkconnection.responses.IResponse;

public abstract class BasePlurkCallback<T extends IResponse> {

    protected abstract void onSuccess(T parsedResponse) throws Exception;

    private NewThreadRetryExecutor.Tasks tasks;

    public void onRetry(Throwable e, long retryTimes, long totalTimes, BuildablePlurkConnection.ErrorAction errorAction) {

    }

    public void onError(Throwable e) {

    }

    public void runResult(Response response) throws Exception{
        if(response.code() != HttpURLConnection.HTTP_OK){
            PlurkConnectionException exception = new PlurkConnectionException(
                    new ApiResponseString(response.code(), response.body().string())
            );
            response.close();
            throw exception;
        }
    }

}

File: library/src/main/java/tw/shounenwind/plurkconnection/callbacks/BasePlurkCallback.kt
package tw.shounenwind.plurkconnection.callbacks

import okhttp3.Response
import tw.shounenwind.plurkconnection.BuildablePlurkConnection
import tw.shounenwind.plurkconnection.PlurkConnectionException
import tw.shounenwind.plurkconnection.Tasks
import tw.shounenwind.plurkconnection.responses.ApiResponseString
import tw.shounenwind.plurkconnection.responses.IResponse
import java.net.HttpURLConnection

abstract class BasePlurkCallback<T : IResponse<*>> {

    private val tasks: Tasks? = null

    @Throws(Exception::class)
    protected abstract fun onSuccess(parsedResponse: T)

    open fun onRetry(e: Throwable, retryTimes: Long, totalTimes: Long, errorAction: BuildablePlurkConnection.ErrorAction) {

    }

    open fun onError(e: Throwable) {

    }

    @Throws(Exception::class)
    open fun runResult(response: Response) {
        if (response.code() != HttpURLConnection.HTTP_OK) {
            val exception = PlurkConnectionException(
                    ApiResponseString(response.code(), response.body()!!.string())
            )
            response.close()
            throw exception
        }
    }

}
